x/cash/dynamicfee_test.go

```diff
@@ -0,0 +1,240 @@
+		handler    *handlerMock
+		minumumFee x.Coin
+		txFee      x.Coin
+		// Wallet state created before running Check
+		initWallets []orm.Object
+		// Wallet state applied after running Check but before running Deliver
+		updateWallets []orm.Object
+
+		wantCheckErr     error
+		wantCheckTxFee   x.Coin
+		wantDeliverErr   error
+		wantDeliverTxFee x.Coin
+		wantGasPayment   int64
+	}{
+		"on success full transaction fee is charged": {
+			signers: []weave.Condition{perm1},
+			handler: &handlerMock{},
+			initWallets: []orm.Object{
+				walletObj(perm1.Address(), 1, 0, "BTC"),
+			},
+			minumumFee:       x.NewCoin(0, 23, "BTC"),
+			txFee:            x.NewCoin(0, 421, "BTC"),
+			wantCheckTxFee:   x.NewCoin(0, 421, "BTC"),
+			wantDeliverTxFee: x.NewCoin(0, 842, "BTC"),
+			wantGasPayment:   421,
+		},
+		"on a handler check failure minumum fee is charged": {
+			signers: []weave.Condition{perm1},
+			handler: &handlerMock{checkErr: ErrTestingError},
+			initWallets: []orm.Object{
+				walletObj(perm1.Address(), 1, 0, "BTC"),
+			},
+			minumumFee:     x.NewCoin(0, 23, "BTC"),
+			txFee:          x.NewCoin(0, 421, "BTC"),
+			wantCheckErr:   ErrTestingError,
+			wantCheckTxFee: x.NewCoin(0, 23, "BTC"),
+		},
+		"on inssuficient fee funds minimum fee is charged": {
+			signers: []weave.Condition{perm1},
+			initWallets: []orm.Object{
+				walletObj(perm1.Address(), 0, 100, "BTC"),
+			},
+			minumumFee:     x.NewCoin(0, 23, "BTC"),
+			txFee:          x.NewCoin(0, 421, "BTC"), // Wallet has not enough.
+			wantCheckErr:   errors.ErrInsufficientAmount,
+			wantCheckTxFee: x.NewCoin(0, 23, "BTC"),
+		},
+		"on inssuficient funds minumum fee withdraw fails": {
+			signers: []weave.Condition{perm1},
+			initWallets: []orm.Object{
+				walletObj(perm1.Address(), 0, 1, "BTC"),
+			},
+			minumumFee:     x.NewCoin(0, 23, "BTC"),  // Wallet has not enough.
+			txFee:          x.NewCoin(0, 421, "BTC"), // Wallet has not enough.
+			wantCheckErr:   errors.ErrInsufficientAmount,
+			wantCheckTxFee: x.Coin{},
+		},
+		"on transaction fee ticker missmatch minimum fee with no currency accepts anything": {
+			signers: []weave.Condition{perm1},
+			initWallets: []orm.Object{
+				walletObj(perm1.Address(), 1, 0, "BTC"),
+			},
+			minumumFee:     x.NewCoin(0, 23, ""),
+			txFee:          x.NewCoin(0, 421, "ETH"),
+			wantCheckErr:   errors.ErrInsufficientAmount,
+			wantCheckTxFee: x.NewCoin(0, 23, "BTC"),
+		},
+		"on a handler deliver failure only minumum fee is charged": {
+			signers: []weave.Condition{perm1},
+			handler: &handlerMock{deliverErr: ErrTestingError},
+			initWallets: []orm.Object{
+				walletObj(perm1.Address(), 1, 0, "BTC"),
+			},
+			minumumFee:       x.NewCoin(0, 11, "BTC"),
+			txFee:            x.NewCoin(0, 44, "BTC"),
+			wantGasPayment:   44, // This assimes that transaction fee was charged.
+			wantCheckTxFee:   x.NewCoin(0, 44, "BTC"),
+			wantDeliverErr:   ErrTestingError,
+			wantDeliverTxFee: x.NewCoin(0, 55, "BTC"),
```

Comment:
I wonder about this case.

This is a fee deducted in total, after successful (no error) `Check` call and `Deliver` call that failed. We charge a mixture of transaction fee and minimum fee (11 + 44 = 55).
The operation was not successful, transaction changes were rolled back. But we charge one full transaction fee and one minimal.

It is strange for me that we charge the transaction fee twice - once for `Check` and once for `Deliver`. This is the same as the static fee decorator though.

Reply:
This is due to wrong usage of the store without a discard. See my other comment. I would expect the minimum fee charged only.